test/images/agnhost/tcp-seq-rst/main.go

```diff
@@ -0,0 +1,188 @@
+	Args:  cobra.MaximumNArgs(0),
+	Run:   main,
+}
+
+var (
+	port int
+)
+
+func init() {
+	CmdTCPSeqRst.Flags().IntVar(&port, "port", 9000, "Port number.")
+}
+
+func main(cmd *cobra.Command, args []string) {
+	// get local IP address
+	ip := getIP()
+	if ip == nil {
+		panic(ip)
+	}
+	log.Printf("external ip: %v", ip.String())
+
+	// listen TCP packets to inject the out of order TCP packets
+	go probe(ip.String())
+
+	log.Printf("listen on %v:%d", "0.0.0.0", port)
+
+	// open a server listening to establish the TCP connections
+	listener, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", port))
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			panic(err)
+		}
+
+		go func(conn net.Conn) {
+			// Close the connection after 10 secs
+			time.Sleep(10 * time.Second)
+			conn.Close()
+		}(conn)
+	}
+}
+
+func probe(ip string) {
+	log.Printf("probing %v", ip)
+
+	ipAddr, err := net.ResolveIPAddr("ip:tcp", ip)
+	if err != nil {
+		panic(err)
+	}
+
+	conn, err := net.ListenIP("ip:tcp", ipAddr)
+	if err != nil {
+		panic(err)
+	}
+
+	// track the pending packets
+	pending := make(map[string]uint32)
+
+	var buffer [4096]byte
+	for {
+		n, addr, err := conn.ReadFrom(buffer[:])
+		if err != nil {
+			log.Printf("conn.ReadFrom() error: %v", err)
+			continue
+		}
+
+		// decode the TCP packet
+		pkt := &tcpPacket{}
+		data, err := pkt.decode(buffer[:n])
+		if err != nil {
+			log.Printf("tcp packet parse error: %v", err)
+			continue
+		}
+
+		// skip unrelated packets
+		if pkt.DestPort != uint16(port) {
+			continue
+		}
+
+		log.Printf("tcp packet: %+v, flag: %v, data: %v, addr: %v", pkt, pkt.FlagString(), data, addr)
+
+		// if is not the first packet increment the sequence
+		if pkt.Flags&SYN != 0 {
+			pending[addr.String()] = pkt.Seq + 1
+			continue
+		}
+
+		// we should not receive a RST because
+		// invalid conntrack packets must be dropped
+		if pkt.Flags&RST != 0 {
+			panic("RST received")
+		}
+
+		// if we are here it means the connection was established
+		if pkt.Flags&ACK != 0 {
+			if seq, ok := pending[addr.String()]; ok {
+				log.Println("connection established")
+				delete(pending, addr.String())
+
+				badPkt := &tcpPacket{
+					SrcPort:    pkt.DestPort,
+					DestPort:   pkt.SrcPort,
+					Ack:        seq,
+					Seq:        pkt.Ack - 100000,      // Bad: seq out-of-window
+					Flags:      (5 << 12) | PSH | ACK, // Offset and Flags  oooo000F FFFFFFFF (o:offset, F:flags)
+					WindowSize: pkt.WindowSize,
+				}
+
+				data := []byte("boom!!!")
+				remoteIP := net.ParseIP(addr.String())
+				localIP := net.ParseIP(conn.LocalAddr().String())
+				_, err := conn.WriteTo(badPkt.encode(localIP, remoteIP, data[:]), addr)
+				if err != nil {
+					log.Printf("conn.WriteTo() error: %v", err)
+				}
+				log.Printf("boom packet: %+v, flag: %v, data: %v, addr: %v", badPkt, badPkt.FlagString(), data, addr)
+
+			}
+		}
+	}
+}
+
+func getIP() net.IP {
```

Comment:
returning a single IP means it will work for single-stack IPv4 or IPv6 but only has a 50-50 chance of doing what you expect in dual-stack. (ie, you might connect via IPv6 but it's snooping IPv4).

Which leads to another problem; if the snooping thread fails catastrophically (eg, snooping the wrong IP), the behavior is that the e2e test will succeed. It would be good if we could verify somehow that it was actually working.